### PR TITLE
Feature/add cache

### PR DESCRIPTION
**Pull request by chrisivens**
Allow the `persistent` option to store the tokens in cache if the grant type is client credentials. This will use the standard PSR-6 interface that Symfony works with and is configured as `cache.app`.

The code in `src/GuzzleBundleOAuth2Plugin.php` which chooses the class to return for the persistence middleware could probably be refactored at some point but does work in our tests against a redis cache storage.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -6,6 +6,11 @@
     "homepage":    "https://github.com/gregurco/GuzzleBundleOAuth2Plugin",
     "license":     "MIT",
 
+    "repositories": [
```
Comment (gregurco) [inline composer.json]: please remove modifications in composer.json

File: src/Middleware/CachedOAuthMiddleware.php [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+     *
+     * @param ClientInterface $client
+     * @param GrantTypeInterface|null $grantType
+     * @param RefreshTokenGrantTypeInterface|null $refreshTokenGrantType
+     * @param AdapterInterface $cacheClient
+     * @param string $clientName
+     */
+    public function __construct(
+        ClientInterface $client,
+        ?GrantTypeInterface $grantType = null,
```
Comment (gregurco) [inline src/Middleware/CachedOAuthMiddleware.php]: plugin supports PHP 7.0, but definition `?type` was introduces in PHP 7.1. Could you check compatibility with 7.0?

File: src/GuzzleBundleOAuth2Plugin.php [hunk 3]
```diff
@@ -90,14 +90,29 @@ public function loadForClient(array $config, ContainerBuilder $container, string
...
-                    $clientName
-                ]);
+                if ($config['grant_type'] === ClientCredentials::class) {
+                    $oAuth2MiddlewareDefinition = new Definition('%guzzle_bundle_oauth2_plugin.cached_middleware.class%');
+                    $oAuth2MiddlewareDefinition->setArguments(
+                        [
+                            new Reference($oauthClientDefinitionName),
+                            new Reference($passwordCredentialsDefinitionName),
+                            new Reference($refreshTokenDefinitionName),
+                            new Reference('Symfony\Component\Cache\Adapter\AdapterInterface'),
```
Comment (gregurco) [inline src/GuzzleBundleOAuth2Plugin.php]: could you write here `new Reference(AdapterInterface::class),` ?

**Review by gregurco — CHANGES_REQUESTED** [review]

File: composer.json (hunk 1 above)
Comment (chrisivens) [reply composer.json]: Sorry about this section, I missed it when prepping the PR.

File: src/Middleware/CachedOAuthMiddleware.php (hunk 2 above)
Comment (chrisivens) [reply src/Middleware/CachedOAuthMiddleware.php]: Done

File: src/GuzzleBundleOAuth2Plugin.php (hunk 3 above)
Comment (chrisivens) [reply src/GuzzleBundleOAuth2Plugin.php]: Not quite sure why I did it the other way, I don't normally use a string literal. Change made.

**Comment by chrisivens** [comment]
Not sure why my commits were not being picked up as I pushed them so I ended up having to force push to get them through.

**Comment by gregurco** [comment]
@chrisivens could you please squash commits to have just one instead of 8? Do you know how to do that?

**Comment by chrisivens** [comment]
Yep, I'll do that.

**Comment by gregurco** [comment]
@chrisivens thank you! Merged :+1:

**Comment by chrisivens** [comment]
Any change of a tagged release for this?

**Comment by gregurco** [comment]
@chrisivens thank you for reminder :slightly_smiling_face: 
Released: https://github.com/gregurco/GuzzleBundleOAuth2Plugin/releases/tag/v1.0.6

**Pull request merged**